File: src/lib/signals.ts
import { existsSync, readFileSync, readdirSync, unlinkSync, mkdirSync } from 'fs';
import { join } from 'path';
import { SIGNALS_DIR } from './paths.js';

export type SessionStatus = 'working' | 'waiting' | 'idle' | 'error';

export interface SessionSignal {
  status: SessionStatus;
  updated: number;
}

export interface WindowStatusInfo {
  text: string;
  status: SessionStatus;
  icon: string;
}

const STATUS_ICONS: Record<SessionStatus, string> = {
  working: '⠿',
  waiting: '◆',
  idle: '○',
  error: '✖',
};

const SIGNAL_MAX_AGE_MS = 5 * 60 * 1000; // 5 minutes

/**
 * Sanitize pane ID to create a safe filename
 */
function sanitizePaneId(paneId: string): string {
  return paneId.replace(/[^a-zA-Z0-9]/g, '_');
}

/**
 * Read signal file for a tmux pane
 */
export function readSignal(paneId: string): SessionSignal | null {
  if (!existsSync(SIGNALS_DIR)) return null;

  const filename = `${sanitizePaneId(paneId)}.json`;
  const filepath = join(SIGNALS_DIR, filename);

  if (!existsSync(filepath)) return null;

  try {
    const content = readFileSync(filepath, 'utf-8');
    const signal = JSON.parse(content) as SessionSignal;

    // Check if signal is stale
    const ageMs = Date.now() - signal.updated * 1000;
    if (ageMs > SIGNAL_MAX_AGE_MS) {
      return null;
    }

    return signal;
  } catch {
    return null;
  }
}

/**
 * Get status icon for a session status
 */
export function getStatusIcon(status: SessionStatus): string {
  return STATUS_ICONS[status];
}

/**
 * Ensure the signals directory exists
 */
export function ensureSignalsDir(): void {
  if (!existsSync(SIGNALS_DIR)) {
    mkdirSync(SIGNALS_DIR, { recursive: true });
  }
}

/**
 * Clean up signal file for a pane
 */
export function cleanupSignal(paneId: string): void {
  const filename = `${sanitizePaneId(paneId)}.json`;
  const filepath = join(SIGNALS_DIR, filename);

  try {
    if (existsSync(filepath)) {
      unlinkSync(filepath);
    }
  } catch {
    // Ignore cleanup errors
  }
}

/**
 * Clean up all stale signal files (older than 1 hour)
 */
export function cleanupStaleSignals(): void {
  if (!existsSync(SIGNALS_DIR)) return;

  const STALE_AGE_MS = 60 * 60 * 1000; // 1 hour
  const now = Date.now();

  try {
    const files = readdirSync(SIGNALS_DIR).filter(f => f.endsWith('.json'));
    for (const file of files) {
      const filepath = join(SIGNALS_DIR, file);
      try {
        const content = readFileSync(filepath, 'utf-8');
        const signal = JSON.parse(content) as SessionSignal;
        const ageMs = now - signal.updated * 1000;
        if (ageMs > STALE_AGE_MS) {
          unlinkSync(filepath);
        }
      } catch {
        // Delete malformed files
        try {
          unlinkSync(filepath);
        } catch {
          // Ignore
        }
      }
    }
  } catch {
    // Ignore errors
  }
}

File: src/signals.rs
use std::fs;
use std::path::PathBuf;

use crate::config;

#[derive(Debug, Clone, PartialEq)]
pub enum SessionStatus {
    Working,
    Waiting,
    Idle,
    Error,
}

#[derive(Debug, Clone)]
pub struct SessionSignal {
    pub status: SessionStatus,
    pub updated: u64,
}

/// How old a signal may be before its status stops meaning anything.
///
/// Shared with `crate::remote_grid`, which applies the same window against the
/// **barn's** clock rather than ours.
pub(crate) const SIGNAL_MAX_AGE_SECS: u64 = 5 * 60; // 5 minutes
const STALE_AGE_SECS: u64 = 60 * 60; // 1 hour

/// Signal files are named after the pane they describe, with everything that is
/// not alphanumeric replaced (`%33` -> `_33`). Shared with `crate::remote_grid`:
/// a barn's filenames arrive already sanitized, so running the same function
/// over a raw pane id makes remote lookups match local ones for free.
pub(crate) fn sanitize_pane_id(pane_id: &str) -> String {
    pane_id
        .chars()
        .map(|c| if c.is_ascii_alphanumeric() { c } else { '_' })
        .collect()
}

fn signal_file(pane_id: &str) -> PathBuf {
    config::signals_dir().join(format!("{}.json", sanitize_pane_id(pane_id)))
}

/// Parse the JSON body of a signal file. No freshness check — the clock to judge
/// against depends on where the signal came from.
///
/// Split out of [`read_signal`] so `crate::remote_grid` can reuse it on a
/// signal that arrived over ssh, where the age has to be measured against the
/// barn's own clock instead of this machine's.
pub(crate) fn parse_signal(content: &str) -> Option<SessionSignal> {
    let json: serde_json::Value = serde_json::from_str(content).ok()?;

    let status_str = json["status"].as_str()?;
    let updated = json["updated"].as_u64()?;

    let status = match status_str {
        "working" => SessionStatus::Working,
        "waiting" => SessionStatus::Waiting,
        "idle" => SessionStatus::Idle,
        "error" => SessionStatus::Error,
        _ => return None,
    };

    Some(SessionSignal { status, updated })
}

/// Read signal file for a tmux pane
pub fn read_signal(pane_id: &str) -> Option<SessionSignal> {
    let dir = config::signals_dir();
    if !dir.exists() {
        return None;
    }

    let filepath = signal_file(pane_id);
    if !filepath.exists() {
        return None;
    }

    let content = fs::read_to_string(&filepath).ok()?;
    let signal = parse_signal(&content)?;

    // Check if signal is stale
    let now = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs();

    if now.saturating_sub(signal.updated) > SIGNAL_MAX_AGE_SECS {
        return None;
    }

    Some(signal)
}

/// Get status icon for a session status
pub fn get_status_icon(status: &SessionStatus) -> &'static str {
    match status {
        SessionStatus::Working => "⠿",
        SessionStatus::Waiting => "◆",
        SessionStatus::Idle => "○",
        SessionStatus::Error => "✖",
    }
}

/// Ensure the signals directory exists
pub fn ensure_signals_dir() {
    let dir = config::signals_dir();
    if !dir.exists() {
        let _ = fs::create_dir_all(&dir);
    }
}

/// Clean up signal file for a pane
pub fn cleanup_signal(pane_id: &str) {
    let filepath = signal_file(pane_id);
    let _ = fs::remove_file(&filepath);
}

/// Clean up all stale signal files (older than 1 hour)
pub fn cleanup_stale_signals() {
    let dir = config::signals_dir();
    if !dir.exists() {
        return;
    }

    let now = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs();

    let entries = match fs::read_dir(&dir) {
        Ok(e) => e,
        Err(_) => return,
    };

    for entry in entries.flatten() {
        let path = entry.path();
        if path.extension().is_some_and(|e| e == "json") {
            let should_delete = match fs::read_to_string(&path) {
                Ok(content) => match serde_json::from_str::<serde_json::Value>(&content) {
                    Ok(json) => match json["updated"].as_u64() {
                        Some(updated) => now.saturating_sub(updated) > STALE_AGE_SECS,
                        None => true, // malformed
                    },
                    Err(_) => true, // malformed
                },
                Err(_) => true,
            };
            if should_delete {
                let _ = fs::remove_file(&path);
            }
        }
    }
}
